archive/imo/imo2019_q2.lean

```diff
@@ -0,0 +1,576 @@
+
+lemma A_ne_C : cfg.A ≠ cfg.C := cfg.affine_independent_ABC.injective.ne
+  (dec_trivial : (0 : fin 3) ≠ 2)
+
+lemma B_ne_C : cfg.B ≠ cfg.C := cfg.affine_independent_ABC.injective.ne
+  (dec_trivial : (1 : fin 3) ≠ 2)
+
+lemma not_collinear_ABC : ¬collinear ℝ ({cfg.A, cfg.B, cfg.C} : set Pt) :=
+affine_independent_iff_not_collinear_set.1 cfg.affine_independent_ABC
+
+lemma wbtw_A_P_A₁ : wbtw ℝ cfg.A cfg.P cfg.A₁ := cfg.P_mem_AA₁
+
+lemma wbtw_A_B₁_C : wbtw ℝ cfg.A cfg.B₁ cfg.C := cfg.B₁_mem_AC
+
+lemma wbtw_B_A₁_C : wbtw ℝ cfg.B cfg.A₁ cfg.C := cfg.A₁_mem_BC
+
+lemma wbtw_B_Q_B₁ : wbtw ℝ cfg.B cfg.Q cfg.B₁ := cfg.Q_mem_BB₁
+
+/-- `ABC` as a `triangle`. -/
+def triangle_ABC : triangle ℝ Pt := ⟨_, cfg.affine_independent_ABC⟩
+
+lemma A_mem_circumsphere : cfg.A ∈ cfg.triangle_ABC.circumsphere :=
+cfg.triangle_ABC.mem_circumsphere 0
+
+lemma B_mem_circumsphere : cfg.B ∈ cfg.triangle_ABC.circumsphere :=
+cfg.triangle_ABC.mem_circumsphere 1
+
+lemma C_mem_circumsphere : cfg.C ∈ cfg.triangle_ABC.circumsphere :=
+cfg.triangle_ABC.mem_circumsphere 2
+
+lemma symm_triangle_ABC : cfg.symm.triangle_ABC = cfg.triangle_ABC.reindex (equiv.swap 0 1) :=
+by { ext i, fin_cases i; refl }
+
+lemma symm_triangle_ABC_circumsphere :
+  cfg.symm.triangle_ABC.circumsphere = cfg.triangle_ABC.circumsphere :=
+by rw [symm_triangle_ABC, affine.simplex.circumsphere_reindex]
+
+/-- `A₂` is the second point of intersection of the ray `AA₁` with the circumcircle of `ABC`. -/
+def A₂ : Pt := cfg.triangle_ABC.circumsphere.second_inter cfg.A (cfg.A₁ -ᵥ cfg.A)
+
+/-- `B₂` is the second point of intersection of the ray `BB₁` with the circumcircle of `ABC`. -/
+def B₂ : Pt := cfg.triangle_ABC.circumsphere.second_inter cfg.B (cfg.B₁ -ᵥ cfg.B)
+
+lemma A₂_mem_circumsphere : cfg.A₂ ∈ cfg.triangle_ABC.circumsphere :=
+(sphere.second_inter_mem _).2 cfg.A_mem_circumsphere
+
+lemma B₂_mem_circumsphere : cfg.B₂ ∈ cfg.triangle_ABC.circumsphere :=
+(sphere.second_inter_mem _).2 cfg.B_mem_circumsphere
+
+lemma symm_A₂ : cfg.symm.A₂ = cfg.B₂ :=
+by { simp_rw [A₂, B₂, symm_triangle_ABC_circumsphere], refl }
+
+lemma QP_parallel_BA : line[ℝ, cfg.Q, cfg.P] ∥ line[ℝ, cfg.B, cfg.A] :=
+by { rw [set.pair_comm cfg.Q, set.pair_comm cfg.B], exact cfg.PQ_parallel_AB }
+
+lemma A_ne_A₁ : cfg.A ≠ cfg.A₁ :=
+begin
+  intro h,
+  have h' := cfg.not_collinear_ABC,
+  rw [h, set.insert_comm] at h',
+  exact h' cfg.wbtw_B_A₁_C.collinear
+end
+
+lemma collinear_PAA₁A₂ : collinear ℝ ({cfg.P, cfg.A, cfg.A₁, cfg.A₂} : set Pt) :=
+begin
+  rw [A₂,
+      (cfg.triangle_ABC.circumsphere.second_inter_collinear cfg.A cfg.A₁).collinear_insert_iff_of_ne
+        (set.mem_insert _ _) (set.mem_insert_of_mem _ (set.mem_insert _ _)) cfg.A_ne_A₁,
+      set.insert_comm],
+  exact cfg.wbtw_A_P_A₁.collinear
+end
+
+lemma A₁_ne_C : cfg.A₁ ≠ cfg.C :=
+begin
+  intro h,
+  have hsbtw := cfg.sbtw_Q_A₁_Q₁,
+  rw h at hsbtw,
+  have ha := hsbtw.angle₂₃₁_eq_zero,
+  rw [angle_CQ₁Q_eq_angle_CBA, angle_eq_zero_iff_ne_and_wbtw] at ha,
```

Comment:
After `rw [angle_CQ₁Q_eq_angle_CBA] at ha` you have a hypothesis `ha: ∠ cfg.C cfg.B cfg.A = 0` and if this doesn't give you a one-line contradiction using `affine_independent_ABC : affine_independent ℝ ![A, B, C]` or `cfg.not_collinear_ABC` then probably you're missing API.

Reply:
I've filed #18069 with some such lemmas and used them to simplify this and another proof, and marked this PR as depending on #18069.